comp/forwarder/defaultforwarder/forwarder_health.go

```diff
@@ -142,6 +152,24 @@ func (fh *forwarderHealth) healthCheckLoop() {
 	}
 }
 
+func (fh *forwarderHealth) updateAPIKey(newAPIKey string) {
+	for domain := range fh.domainResolvers {
+		if domainURLRegexp.MatchString(domain) {
+			domain = "https://api." + domainURLRegexp.FindString(domain)
+		}
+		prevKeys := fh.keysPerAPIEndpoint[domain]
+		replaceKeys := make([]string, 0, len(prevKeys))
+		for _, oldKey := range prevKeys {
+			replaceKeys = append(replaceKeys, newAPIKey)
+			// remove both success and failure messages
+			fh.setAPIKeyStatus(oldKey, "", &apiKeyRemove)
+		}
+		fh.keysPerAPIEndpoint[domain] = replaceKeys
+	}
+	// check apiKey validity and update the messages
+	fh.checkValidAPIKey()
+}
```

Comment:
<!-- Warning represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 warning:** ‏Not entirely sure but I guess this is called concurrently with the usual forwarder work, so we might need some synchronization around `fh.keysPerAPIEndpoint` ? (ie. a `RWLock`)

Reply:
After you mentioned mutexs I realized that there's memory aliasing going on between this and the defaultForwarder. Solving it involves a non-trivial refactor, but I'm working on it today.

Reply:
Nevermind, this was simpler than I thought. Added a mutex around the `fh.keysPerAPIEndpoint` as you said, and added mutexes to the `DomainResolvers` so that they can update and get their API keys safely.